File: problems/35-string-dp/003-longest-palindromic-subsequence/java/Solution.java
public class Solution {
    public int longestPalindromeSubseq(String s) {
        int n=s.length(); if(n==0)return 0;
        int[][] dp=new int[n][n];
        for(int i=n-1;i>=0;i--){dp[i][i]=1; for(int j=i+1;j<n;j++) dp[i][j]=s.charAt(i)==s.charAt(j)?2+(i+1<=j-1?dp[i+1][j-1]:0):Math.max(dp[i+1][j],dp[i][j-1]);}
        return dp[0][n-1];
    }
}

File: problems/35-string-dp/003-longest-palindromic-subsequence/python/solution.py
class Solution:
    def longest_palindrome_subseq(self, s):
        n=len(s); dp=[[0]*n for _ in range(n)]
        for i in range(n-1,-1,-1):
            dp[i][i]=1
            for j in range(i+1,n):
                dp[i][j]=2+(dp[i+1][j-1] if i+1<=j-1 else 0) if s[i]==s[j] else max(dp[i+1][j],dp[i][j-1])
        return dp[0][n-1] if n else 0
